### PR TITLE
Adds option to edit which file extensions are searched

### LINKED ISSUE #114: Nested jars: extension settings
Add frontend settings to show default extensions used and allow for their addition/removal.

### PR DESCRIPTION
**Pull request by ilbinek**
Adds a window to edit which file extensions are used while searching though nested jars.

Keep in mind that listing of classes is not influenced by this, as it uses different method of iterating though the nested jars.

Fixes #114 

image
image

### CODE REVIEW

File: runtime-decompiler/src/main/java/org/jrd/frontend/frame/main/MainFrameView.java
Comment (judovana) [inline runtime-decompiler/src/main/java/org/jrd/frontend/frame/main/MainFrameView.java:318]: Lets introduce general "setting" menu item here - and put tabbed pane in. NestedJars will be one tab. class path less compiler should be another, and agent setup should be third. For plugin, the separate menu item is still worthy.

Agents have to remian first tab.

Thanx!

File: runtime-decompiler/src/main/java/org/jrd/frontend/frame/nestedJars/NestedJarsView.java
Comment (judovana) [inline runtime-decompiler/src/main/java/org/jrd/frontend/frame/nestedJars/NestedJarsView.java:45]: nonoon. Do not enforce the dot please. Abit of freedom would be nice :)

File: runtime-decompiler/src/main/java/org/jrd/frontend/frame/nestedJars/NestedJarsView.java
Comment (judovana) [inline runtime-decompiler/src/main/java/org/jrd/frontend/frame/nestedJars/NestedJarsView.java:55]: pls use as tooltip, ot better as something waht can be copypasted, to show current list of default extensions

File: runtime-decompiler/src/main/java/org/jrd/frontend/frame/nestedJars/NestedJarsView.java
Comment (judovana) [inline runtime-decompiler/src/main/java/org/jrd/frontend/frame/nestedJars/NestedJarsView.java:35]: I'm in temptation to get rid of jlist and add/remove buttons, in favour of single textarea (multilined, split is done on \s+ with newLines counted anyway).
Nice gui havee jidea - you see it as list, but once you click on edit, it changes to textfield, so yo just write.
Imagine, to write "ext1 ext2 extt3....extN" enter/ok is much better then click "ext1" click, click "ext2" click, .... click extN click...
Up to yoiu

Comment (ilbinek) [reply runtime-decompiler/src/main/java/org/jrd/frontend/frame/nestedJars/NestedJarsView.java:35]: I'll add check that will split the textfield text by " " and use them as different extensions

File: runtime-decompiler/src/main/java/org/jrd/frontend/frame/nestedJars/NestedJarsView.java
Comment (ilbinek) [reply runtime-decompiler/src/main/java/org/jrd/frontend/frame/nestedJars/NestedJarsView.java:58]: Tooltip added. It could also be added into the text, let me know if that's something you'd like.

**Comment by judovana** [comment]
"Multiple suffixes can be added at once. "  It is not the same user experience but may be enough.
Pls split on \s+ not just " "

File: runtime-decompiler/src/main/java/org/jrd/frontend/frame/main/MainFrameView.java
Comment (AurumTheEnd) [reply runtime-decompiler/src/main/java/org/jrd/frontend/frame/main/MainFrameView.java:318]: I would be of the opinion to put of the mentioned settings into what is now called `Configure -> Agent` (should be renamed to `Configure -> Settings`). There is plenty of room in that window, and making all the settings visible at once is a benefit.

Comment (judovana) [reply runtime-decompiler/src/main/java/org/jrd/frontend/frame/main/MainFrameView.java:318]: agent may arrive any time later.

File: runtime-decompiler/src/main/java/org/jrd/frontend/frame/settings/SettingsView.java
Comment (AurumTheEnd) [inline runtime-decompiler/src/main/java/org/jrd/frontend/frame/settings/SettingsView.java:203]: Suggestion for better readability:

```java
List<String> extensions;
if (settingsPanel.useDefaults.isSelected()) {
    extensions = new ArrayList<String>();
} else {
    extensions = Collections.list(settingsPanel.dList.elements());
}
ArchiveManagerOptions.getInstance().setExtension(extensions);
```

File: runtime-decompiler/src/main/java/org/jrd/frontend/frame/settings/SettingsView.java
Comment (AurumTheEnd) [inline runtime-decompiler/src/main/java/org/jrd/frontend/frame/settings/SettingsView.java:145]: You are using one less condition for determining the usage of defaults than in `ArchiveManagerOptions#isInner()`. For this confusion alone, I would recommend to extract that check into a method in `ArchiveManagerOptions` and to use that.

**Review by AurumTheEnd — CHANGES_REQUESTED** [review]

File: runtime-decompiler/src/main/java/org/jrd/frontend/frame/settings/SettingsView.java
Comment (ilbinek) [reply runtime-decompiler/src/main/java/org/jrd/frontend/frame/settings/SettingsView.java:115]: "Calling this method with a negative value results in the property being set to 0."

This is how I found it used in the example I was going through, if you wish it to be changed to 0, I have no problem with it.

**Comment by AurumTheEnd** [comment]
Couple things needed to merge this:
 - Please resolve the conversations which you have corrected. At a quick glance, not everything has been fixed yet. This is needed so that I can do a (hopefully) final review before merging.
 - Please resolve the merge conflicts caused by #128. One of the things changed in it was rewriting a nested jar related class from tab indentations to space indentations, see if that causes you any problems.
 - Please rebase the correction commits into the original commits. Again, this is standard practice, because you are fixing what you yourself introduced as a problem. I will not be squashing this PR, so the correction commits have no business being standalone after the merge == on the master branch.
No idea how you would do this in your Git GUI, but on the terminal you'd do something like `git rebase -i OLDEST_COMMIT_SHA^` and work with the commits from there.
Note that you don't need to include the correction commit messages in the oldest commit - after the PR merge, they won't matter!

File: runtime-decompiler/src/main/java/org/jrd/frontend/frame/settings/SettingsView.java
Comment (AurumTheEnd) [inline runtime-decompiler/src/main/java/org/jrd/frontend/frame/settings/SettingsView.java]: Unfortunate to see it go but understandable. No changes needed.

File: runtime-decompiler/src/main/java/org/jrd/frontend/frame/settings/SettingsView.java
Comment (AurumTheEnd) [inline runtime-decompiler/src/main/java/org/jrd/frontend/frame/settings/SettingsView.java:107]: See 
```java
.addKeyListener(new KeyAdapter() {
            @Override
            public void keyReleased(KeyEvent e) {
```
 in BytecodeDecompilerView. Please apply that knowledge (your own changeset from March btw lol) here.

File: runtime-decompiler/src/main/java/org/jrd/frontend/frame/settings/SettingsView.java
Comment (AurumTheEnd) [inline runtime-decompiler/src/main/java/org/jrd/frontend/frame/settings/SettingsView.java:122]: When nothing is selected, this prints an ugly trace. Please catch the exception and log it appropriately.

File: runtime-decompiler/src/main/java/org/jrd/frontend/frame/settings/SettingsView.java
Comment (AurumTheEnd) [inline runtime-decompiler/src/main/java/org/jrd/frontend/frame/settings/SettingsView.java:135]: `getExtensions();` returns not only default extensions, but also currently set extensions. Therefore, something like 'currentExtensions' would be more proper.

**Pull request merged**
